crates/sui-bridge-indexer/src/eth_bridge_indexer.rs

```diff
@@ -175,87 +183,151 @@ impl Datasource<RawEthData> for EthFinalizedSyncDatasource {
+                    target_checkpoint,
+                    data_sender,
+                    indexer_metrics,
+                )
+                .await?;
+            }
+            Ok::<_, Error>(())
+        });
 
-            let mut data = Vec::new();
-            let mut first_block = 0;
+        Ok(handle)
+    }
+}
 
-            for log in logs {
-                let block = if let Some(cached_block) = cached_blocks.get(&log.block_number) {
-                    cached_block.clone()
-                } else {
-                    let Ok(Ok(Some(block))) = retry_with_max_elapsed_time!(
-                        provider.get_block(log.block_number),
-                        Duration::from_secs(30000)
-                    ) else {
-                        panic!("Unable to get block from provider");
-                    };
+async fn retrieve_and_process_live_finalized_logs(
+    client: Arc<EthClient<MeteredEthHttpProvier>>,
+    provider: Arc<Provider<Http>>,
+    addresses: Vec<EthAddress>,
+    starting_checkpoint: u64,
+    data_sender: DataSender<RawEthData>,
+    indexer_metrics: BridgeIndexerMetrics,
+    bridge_metrics: Arc<BridgeMetrics>,
+) {
+    let eth_contracts_to_watch = HashMap::from_iter(
+        addresses
+            .iter()
+            .map(|address| (*address, starting_checkpoint)),
+    );
+
+    let (_, mut eth_events_rx, _) = EthSyncer::new(client.clone(), eth_contracts_to_watch)
+        .run(bridge_metrics.clone())
+        .await
+        .expect("Failed to start eth syncer");
+
+    // forward received events to the data sender
+    while let Some((_, block, logs)) = eth_events_rx.recv().await {
+        let raw_logs: Vec<RawEthLog> = logs
+            .into_iter()
+            .map(|log| RawEthLog {
+                block_number: block,  // Handle block number conversion
+                tx_hash: log.tx_hash, // Assuming `event` contains `tx_hash`
+                log: log.log,
+            })
+            .collect();
+
+        process_logs(raw_logs, provider.clone(), data_sender.clone(), block)
+            .await
+            .expect("Failed to process logs");
+        indexer_metrics.latest_committed_eth_block.set(block as i64);
+    }
 
-                    cached_blocks.insert(log.block_number, block.clone());
-                    block
-                };
+    panic!("Eth syncer stopped unexpectedly");
+}
 
-                if first_block == 0 {
-                    first_block = log.block_number;
-                }
+async fn retrieve_and_process_log_range(
+    client: Arc<EthClient<MeteredEthHttpProvier>>,
+    provider: Arc<Provider<Http>>,
+    addresses: Vec<EthAddress>,
+    starting_checkpoint: u64,
+    target_checkpoint: u64,
+    data_sender: DataSender<RawEthData>,
+    indexer_metrics: BridgeIndexerMetrics,
+) -> Result<(), Error> {
+    let Ok(Ok(logs)) = retry_with_max_elapsed_time!(
+        client.get_raw_events_in_range(addresses.clone(), starting_checkpoint, target_checkpoint),
```

Comment:
`fn get_raw_events_in_ranges` has this comment:
```
    // Note: query may fail if range is too big. Callsite is responsible
    // for chunking the query.
```

We need to segment it (e.g. 1k blocks per chunk) before calling `get_raw_events_in_range`